Basic.wl

```diff
@@ -180,10 +161,36 @@ ImprisonmentExpressions[template_List]:= Cases[template, x_ \[Element] set_ ,Inf
 ValueRestrictions[imprisonmentExpression_]:=
  Apply[Or,imprisonmentExpression[[1]] == #&/@ imprisonmentExpression[[2]]];
 
+RemoveContainedSubsSets[allSubSet_List, subSet_List] :=
+ Module[{listToRemove, result},
+  listToRemove = ContainsAll[#, subSet] && subSet =!= # & /@ allSubSet;
+  result = MapThread[If[! #2, #1, ## &[]] &, {allSubSet, listToRemove}]
+];
+
+BooleanFromRule[x_ /; x[[2]] == 0] := Not[x[[1]]];
+BooleanFromRule[x_ /; x[[2]] == 1] := x[[1]];
+
+ConjunctionFromRuleSet[ruleSet_List] :=
+    Apply[And, Map[BooleanFromRule, ruleSet]];
+
+DNFFromRuleSet[ruleSet_List] :=
+    Apply[Or, Map[ConjunctionFromRuleSet, ruleSet]];
+
+MinimizedRuleSets[ruleSets_List] :=
+    RuleSetFromDNFF[BooleanMinimize[DNFFromRuleSet[ruleSets]]];
+
+RuleSetFromDNFF[minimizeBooleanExpression_] := Module[{RootExpressionToSubSet, ExpressionToSubSet, LeafExpressionToSubSet},
+  RootExpressionToSubSet[expr_ /; Head[expr] === Or] := Map[ExpressionToSubSet, expr, {1}] /. And -> List /. Or -> List;
+  RootExpressionToSubSet[expr_] := {ExpressionToSubSet[expr]} /. And -> List /. Or -> List;
+
+  ExpressionToSubSet[expr_ /; Head[expr] === And] := Map[LeafExpressionToSubSet, expr];
+  ExpressionToSubSet[expr_] := {LeafExpressionToSubSet[expr]};
 
-ExceptionTemplates[intemplate_, k_Integer:2, r_Integer:1] :=
-  MapThread[If[#2=== _,#1,#2]&,{OldBaseTemplate[k,r],#}]&/@Union[(If[NumberQ[#],#,_]&/@#)&/@((OldBaseTemplate[k,r]/.#[[1]])&/@Cases[{#[[2]],#[[1]]/.#[[2]]}&/@Flatten[Outer[List,{#[[1]]},#[[2]],1]&/@({#[[2]],MapThread[#1->#2&,{#[[1]],#[[2]]}]&/@#[[1]]}&/@({First@Outer[List,{#[[1]]},#[[2]],1],#[[3]]}&/@({#[[1]],Tuples[Range[0,k-1],Length[#[[1]]]],#[[2]]}&/@({RuleTemplateVars[{#}],#}&/@Select[intemplate,(Depth[#]>1)&])))),2],{_,x_/;\[Not]MemberQ[Range[0,k-1],x]}])]
+  LeafExpressionToSubSet[expr_ /; Head[expr] === Symbol] := expr -> 1;
```

Comment:
Also, if `ExpressionToSubSet` is not being used, then `LeafExpressionToSubSet` is also not being used.